fractions/microprofile/microprofile-jwt/src/main/java/org/wildfly/swarm/microprofile/jwtauth/deployment/principal/JWTCallerPrincipalFactory.java

```diff
@@ -86,17 +86,17 @@ private static JWTCallerPrincipalFactory loadSpi(ClassLoader cl) {
             try {
                 for (JWTCallerPrincipalFactory spi : sl) {
                     if (instance != null) {
-                        throw new IllegalStateException(
-                                "Multiple JWTCallerPrincipalFactory implementations found: "
-                                        + spi.getClass().getName() + " and "
-                                        + instance.getClass().getName());
+                        throw new MultipleFactoriesException(spi.getClass().getName());
                     } else {
                         log.debugf("sl=%s, loaded=%s", sl, spi);
                         instance = spi;
                     }
                 }
+            } catch (MultipleFactoriesException e) {
+                log.warn("Multiple JWTCallerPrincipalFactory implementations found: "
+                          + e.getSpiClassName() + " and " + instance.getClass().getName());
             } catch (Throwable e) {
-                log.warn("Failed to locate JWTCallerPrincipalFactory provider", e);
+                log.debugf("Failed to locate JWTCallerPrincipalFactory provider, e=%s", e);
```

Comment:
I believe the `java.util.ServiceLoader` is implemented lazily. So if `ServiceLoader` fails to load a service provider (during the first iteration) the information will be lost unless you have the debug level enabled.

Reply:
Thanks, good catch; hmm, I wonder if I should update the PR to clone the iterator first, before entering the try section, otherwise, without rewriting the loading sequence, it is hard to find out in the catch if it was a cast issue related to the use of the parent loader or not; let me try...